DOC: update scipy.integrate section (#586)
scipy-lectures/scipy-lecture-notes: intro/scipy/index.rst

Numerical integration: :mod:`scipy.integrate`
---------------------------------------------

Quadrature
..........

Suppose we wish to compute the definite integral

> :math:`\int_0^{\pi / 2} sin(t) dt` numerically. :func:`scipy.integrate.quad`

Author: ```suggestion
:math:`\int_0^{\pi / 2} \sin(t) dt` numerically. :func:`scipy.integrate.quad`
```

chooses one of several adaptive techniques depending on the parameters, and

> therefore is a reasonable first choice for integration of a single variable::

Author: ```suggestion
is therefore the recommended first choice for integration of a single variable::
```

    >>> integral, error_estimate = sp.integrate.quad(np.sin, 0, np.pi/2)
    >>> np.allclose(integral, 1)  # numerical result ~ analytical result
    True
    >>> abs(integral - 1) < error_estimate  #  actual error < estimated error
    True

Other functions for *numerical quadrature*, including integration of
multivariate functions and approximating integrals from samples, are available
in :mod:`scipy.integrate`.

Initial Value Problems
......................

:mod:`scipy.integrate` also features routines for integrating `Ordinary
Differential Equations (ODE)
<https://en.wikipedia.org/wiki/Ordinary_differential_equation>`__.
For example, :func:`scipy.integrate.solve_ivp` integrates ODEs of the form::

.. math::

    \frac{dy}{dt} = f(t, y(t))

from an initial time :math:`t_0` and initial state :math:`y(t=t_0)=t_0` to a final
time :math:`t_f` or until an event occurs (e.g. a specified state is reached).

As an introduction, consider the initial value problem given by
:math:`\frac{dy}{dt} = -2 y` and the initial condition :math:`y(t=0) = 1` on
the interval :math:`t = 0 \dots 4`. We begin by defining a callable that

> computes the :math:`f(t, y(t))` given the current time and state.

Author: ```suggestion
computes :math:`f(t, y(t))` given the current time and state.
```

    >>> def f(t, y):
    ...     return -2 * y

Then, to compute ``y`` as a function of time::

    >>> t_span = (0, 4)  # time interval
    >>> t_eval = np.linspace(*t_span)  # times at which to evaluate `y`
    >>> y0 = [1,]  # initial state
    >>> res = sp.integrate.solve_ivp(f, t_span=t_span, y0=y0, t_eval=t_eval)

and plot the result::

    >>> plt.plot(res.t, res.y[0])  # doctest: +SKIP
    >>> plt.xlabel('t')  # doctest: +SKIP
    >>> plt.ylabel('y')  # doctest: +SKIP
    >>> plt.title('Solution of Initial Value Problem')  # doctest: +SKIP

.. image:: auto_examples/images/sphx_glr_plot_solve_ivp_simple_001.png
    :target: auto_examples/plot_solve_ivp_simple.html
    :scale: 70
    :align: right

Let us integrate a more complex ODE: a `damped
spring-mass oscillator
<https://en.wikipedia.org/wiki/Harmonic_oscillator#Damped_harmonic_oscillator>`__.
The position of a mass attached to a spring obeys the 2nd order ODE

> :math:`\ddot{y} + 2 \zeta \omega_0  \dot{y} + \omega_0^2 y = 0` with natural frequency

Author: Use of $\zeta$ for the [damping ratio](https://en.wikipedia.org/wiki/Damping) is nearly universal; I've not seen $\varepsilon$ used aside from here.

The position of a mass attached to a spring obeys the 2nd order *ODE*
:math:`y'' + 2 \varepsilon \omega_0  y' + \omega_0^2 y = 0` with
:math:`\omega_0^2 = k/m` with :math:`k` the spring constant, :math:`m` the mass
and :math:`\varepsilon = c/(2 m \omega_0)` with :math:`c` the damping coefficient. We set::
If we define :math:`z = [z_0, z_1]` where :math:`z_0 = y` and :math:`z_1 = \dot{y}`,
then the first order equation:

    >>> mass = 0.5  # kg
    >>> kspring = 4  # N/m
    >>> cviscous = 0.4  # N s/m
.. math::

Hence::
    \usepackage{amsmath}
    \begin{eqnarray}
    \frac{dz}{dt} =
    \begin{bmatrix}
        \frac{dz_0}{dt} \\
        \frac{dz_1}{dt}
    \end{bmatrix} =
    \begin{bmatrix}
        z_1  \\
        -(2 \zeta \omega_0  z_1 + \omega_0^2 z_0)
    \end{bmatrix}
    \end{eqnarray}

    >>> eps = cviscous / (2 * mass * np.sqrt(kspring/mass))
    >>> omega = np.sqrt(kspring / mass)
is equivalent to the original second order equation.

The system is underdamped, as::
We set::

    >>> eps < 1
    True
    >>> m = 0.5  # kg
    >>> k = 4  # N/m
    >>> c = 0.4  # N s/m
    >>> zeta = c / (2 * m * np.sqrt(k/m))
    >>> omega = np.sqrt(k / m)

> For :func:`~scipy.integrate.solve_ivp`, the 2nd order equation

Author: This is often tricky for those who are unfamiliar with the process. I tried to make it more explicit.

>     ...     return (z[1], -2.0 * zeta * omega * z[1] - omega **2 * z[0])

Author: ```suggestion
    ...     return (z[1], -2.0 * zeta * omega * z[1] - omega**2 * z[0])
```

Integration of the system follows::

    >>> t_span = (0, 10)
    >>> t_eval = np.linspace(*t_span, 100)
    >>> z0 = [1, 0]

>     >>> res = sp.integrate.solve_ivp(f, t_span, z0, args=(zeta, omega),
>     ...                              method='LSODA')

Author: ```suggestion
    >>> res = sp.integrate.solve_ivp(f, t_span, z0, t_eval=t_eval,
    ...                              args=(zeta, omega), method='LSODA')
```
